sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -958,6 +958,26 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
 
       project.copy(child = Filter(replaceAlias(condition, aliasMap), grandChild))
 
+    // Push [[Filter]] operators through [[Window]] operators. Parts of the predicate that can be
+    // pushed beneath must satisfy three conditions:
+    // 1. involving one and only one column that is part of window partitioning key.
```

Comment:
So I am not sure I agree with you here.

Lets take your example. We filter the rows  with the following predicate: `key + value > 2`. We use both `key` and `value` in `PARTITION BY` clause, this means they are constant during window function evaluation. The value of `key + value > 1` will (as a result) also be constant during window evaluation. This predicate would filter out entire partitions, so we can safely push it down.

I think we can safely push down any deterministic filter which only references partitioning columns. Let me know what you think.

I am not sure why Hive does not push this down, but this could well be due to the way Hive evaluates window functions (PTFs).